src/ARMInterpreter.cpp

```diff
@@ -220,26 +221,46 @@ void A_MRC(ARM* cpu)
 
 void A_SVC(ARM* cpu)
 {
-    u32 oldcpsr = cpu->CPSR;
-    cpu->CPSR &= ~0xBF;
-    cpu->CPSR |= 0x93;
-    cpu->UpdateMode(oldcpsr, cpu->CPSR);
-
-    cpu->R_SVC[2] = oldcpsr;
-    cpu->R[14] = cpu->R[15] - 4;
-    cpu->JumpTo(cpu->ExceptionBase + 0x08);
+    // Check if BIOS is loaded.
+    if (cpu->useHLE_bios)
+    {
+        u32 swinum = (cpu->CurInstr >> 16) & 0x1F;
+        s32 cycles = ARM_swi_tab[cpu->Num][swinum](cpu);
+        cpu->Cycles += cycles + 3;
+    }
+    else
+    {
+        u32 oldcpsr = cpu->CPSR;
+        cpu->CPSR &= ~0xBF;
+        cpu->CPSR |= 0x93;
+        cpu->UpdateMode(oldcpsr, cpu->CPSR);
+
+        cpu->R_SVC[2] = oldcpsr;
+        cpu->R[14] = cpu->R[15] - 4;
+        cpu->JumpTo(cpu->ExceptionBase + 0x08);
+    }
 }
 
 void T_SVC(ARM* cpu)
 {
-    u32 oldcpsr = cpu->CPSR;
-    cpu->CPSR &= ~0xBF;
-    cpu->CPSR |= 0x93;
-    cpu->UpdateMode(oldcpsr, cpu->CPSR);
-
-    cpu->R_SVC[2] = oldcpsr;
-    cpu->R[14] = cpu->R[15] - 2;
-    cpu->JumpTo(cpu->ExceptionBase + 0x08);
+    // Check if BIOS is loaded.
+    if (cpu->useHLE_bios)
+    {
+        u32 swinum = cpu->CurInstr & 0x1F;
+        s32 cycles = ARM_swi_tab[cpu->Num][swinum](cpu);
+        cpu->Cycles += cycles + 3;
+    }
+    else
+    {
+        u32 oldcpsr = cpu->CPSR;
+        cpu->CPSR &= ~0xBF;
+        cpu->CPSR |= 0x93;
+        cpu->UpdateMode(oldcpsr, cpu->CPSR);
+
+        cpu->R_SVC[2] = oldcpsr;
+        cpu->R[14] = cpu->R[15] - 2;
+        cpu->JumpTo(cpu->ExceptionBase + 0x08);
+    }
```

Comment:
Can copy-paste be avoided?

Reply:
If you mean A_SVC and T_SVC, they were already separate and neither the original or thew new HLE code blocks are identical. (Line 249 has no >>1; Line 261 ends with -2.) I didn't figure it was worth refactoring at this point, and messing with the code structure pattern.